### ISSUE TITLE
[Feature Request] AlOpsCompile - Support for CustomCodeCops

### ISSUE THREAD

**Issue by fvet** [HodorNV/ALOps#430]
**Is your feature request related to a problem? Please describe.**

We'd search some support to add **BusinessCentral.LinterCop.dll** to our pipelines (https://github.com/StefanMaron/BusinessCentral.LinterCop)

- How can we pass the -CustomCodeCops to the ALC (AlOpsCompile) step. 
- Where to put this file? How does ALOps makes sure to get the latest version of the dll?

![image](https://user-images.githubusercontent.com/25268332/134909379-ef52940f-a3e0-46a7-8a6f-33a18af72651.png)

**Comment by waldo1001** [comment]
Because of the way MS implemented this, it isn't supported yet for ALOps (the fact the "analyzerfolder" needs to be added makes it that ALOps doesn't find anything).  

But it's on the roadmap since the LinterCop.

**Comment by fvet** [comment]
Any ETA for this feature? 
BCLinterCop has some recent new rules we'd like to integrate in our pipelines ;)

**Comment by waldo1001** [comment]
The problem with the lintercop is that it needs a fixed reference to a certain version of the alc.  Since we use the compiler of the artifact, it hardly every is 100% the same as the compiled reference.

That's what we're struggling with at the moment ..

**Comment by waldo1001** [comment]
We're building our own custom codecop to see how we can mitigate this issue - and we will inform Stefan when we have a solution .. so we need a bit more time.

**Comment by fvet** [comment]
@waldo1001 Any updated ETA on this?

**Comment by fvet** [comment]
@waldo1001 Would it be possible on short term - untill ALOPS provides custom codecop support - to temporarily exclude the custom cops / Lintercop from the settings.json in the ALOpsCompile task? This way, we could choose to enable the LinterCop on our local dev boxes (by including it in the settings.json) and when ALOps would compile the app, it should exclude all but the default code analyzers, so it does not fail on non-available cops?

**Comment by waldo1001** [comment]
We indeed didn't figure out yet how to work with it - how the custom cops are built is that they are very dependent on the exact version of the microsoft-dlls - which we need to find a solution for to include them while compiling.  We do have a repro for this .. it's a matter of time.

But for short term, this is a good idea: we'll add a setting like "ExcludeCustomCodeCops".

**Comment by fvet** [comment]
@waldo1001 Concerning the short term solution. Any chance in having this included in the next (February?) release?

**Comment by jcoe-nvt** [comment]
@waldo1001 Any ETA on the short term solution?

**Comment by dsaveyn** [comment]
@waldo1001 Were are also running into this issue. Any ETA on this?

**Comment by waldo1001** [comment]
Sorry - there hasn't been any progress to a solution.  Would have loved the codecop-architecture to behave differently :(.  
Still searching when we have time.

**Comment by Arthurvdv** [comment]
Just thinking out loud and unrestrained without any kind of knowledge; Can we compile the LinterCop.dll in the pipeline after retrieving the artifact?

This would be a temporary ALOps building block while we wait for different solution in the unknown future. The step would be to download the source code from the LinterCop and compile against the dll from the artifact, so we have a match. This could be generic to a certain point. Any custom  codecop which you have access to the source code could work?

**Comment by waldo1001** [comment]
Just FYI
we're still considering this - and will now try to mitigate it differently with another approaches:
- download the latest marketplace vsix
- compile on the fly

**Comment by jcoe-nvt** [comment]
We achieved compiling with LinterCop by: 
- Adding a custom step to download LinterCop (see [script](https://github.com/StefanMaron/vsc-lintercop/blob/master/DownloadFile.ps1)) into our working folder
- Setting alternativevsixurl to Latest in Compile step

**Comment by waldo1001** [comment]
Hm .. interesting.  Didn't know of that script.  May be it's a viable solution afterall (as long as Stefan keeps his cop updated to the latest dlls from MS).  

Thanks so much!

**Comment by Arthurvdv** [comment]
> We achieved compiling with LinterCop by:
> 
> * Adding a custom step to download LinterCop (see [script](https://github.com/StefanMaron/vsc-lintercop/blob/master/DownloadFile.ps1)) into our working folder
> * Setting alternativevsixurl to Latest in Compile step

This seemed like a feasible solution, so I've try'ed to implement this in our pipelines.
Unfortunately I'm stuck on the version differences between AL and LinterCop.

```
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0001FlowFieldsShouldNotBeEditable cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0002CommitMustBeExplainedByComment cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0003DoNotUseObjectIDsInVariablesOrProperties cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0004LookupPageIdAndDrillDownPageId cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0005VariableCasingShouldNotDIfferFromDeclaration cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0006FieldNotAutoIncrementInTemporaryTable cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0007DataPerCompanyShouldAlwaysBeSet cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0008NoFilterOperatorsInSetRange cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0009CodeMetrics cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0011DataPerCompanyShouldAlwaysBeSet cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0012DoNotUseObjectIdInSystemFunctions cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0013CheckForNotBlankOnSingleFieldPrimaryKeys cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0014PermissionSetCaptionLength cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0015PermissionSetCoverage cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0016CheckForMissingCaptions cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
##[warning]warning AL1003: An instance of analyzer BusinessCentral.LinterCop.Design.Rule0017WriteToFlowField cannot be created from C:\azure-vsts-agent-build\_work\345\s\BusinessCentral.LinterCop.dll : Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=9.0.9.15366, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
```
![image](https://user-images.githubusercontent.com/44637996/165556140-c35e2e31-5af0-4ad6-a74a-17db569ab2d0.png)
I've noticed when setting the alternativevsixurl to '9.0.605172' it indeed works. But the versions higher then this and/or Latest (= v9.0.615906) throws the error of above.

Theoretically I could manually change the alternativevsixurl to a matching value of the BC LinterCop, but that's way to sensitive to errors.

What I can't wrap my mind around, is that the VS Code on my laptop is installed with the AL Language 'v9.0.615906' and the BC LinterCop 0.25.1 and this combination doesn't thrown an error and works? @jcoe-nvt maybe any idea what I am missing here?

**Comment by waldo1001** [comment]
The best way to make this work is to make sure that everything is always on the latest.  
We recently updated our own codecop to automatically update every day.  I will communicate this practice to Stefan - hopefully he'll schedule the same...

**Comment by jcoe-nvt** [comment]
@Arthurvdv Currently there seem to be mismatches in the latest vsix and lintercop. I had to disable LinterCop to stop my pipelines from failing on these warnings. I noticed this issue on 21-4-2022 for the first time (so the day after new vsix was released). 
Hopefully Stefan will update LinterCop soon (and keep up with the latest versions instantly)

**Comment by jcoe-nvt** [comment]
@Arthurvdv Issue has been solved by Stefan and he implemented the suggestion by @waldo1001 to prevent future issues

**Comment by fvet** [comment]
@waldo1001  Any plans on whether ALOps will support the BCLinterCop on short-term?
Or should we revert to custom solutions? 

@jcoe-nvt @Arthurvdv  Any script to share ;) ?

**Comment by jcoe-nvt** [comment]
@fvet I just use this script from Stefan to download LinterCop in our working directory: [https://github.com/StefanMaron/vsc-lintercop/blob/master/DownloadFile.ps1](https://github.com/StefanMaron/vsc-lintercop/blob/master/DownloadFile.ps1)
And alternativevsixurl needs to be set to Latest in your compile step.

I consider this solution as a support for Custom CodeCops. Previously if analyzer wasn't found in the analyzer folder of the vsix, then your pipeline would give warnings about missing dll. Now, it will fall back to your working directory to see if code cop can be found there.

This way you can not only include LinterCop, but any custom codecop you would like to add to your app.

The compatibly between lintercop and the vsix version isn't something @waldo1001 can do much about (That's dependend on Stefan)

**Comment by waldo1001** [comment]
Indeed.  All we can do is to support for custom codecops .. we can't control if they compile at the right time.  But I just talked to Stefan, and he implemented a daily check.

All we need to do is use the latest VSIX by using `alternativevsixurl: Latest`

**Comment by waldo1001** [comment]
I am prepping an example, though ..

**Comment by jcoe-nvt** [comment]
What you might need to do if you want to compile your app against next minor/major is removing lintercop from your settings.json. This way you can use the vsix that corresponds with this next version and see the new warnings that will be added in the standard codecops in next release(s)

This can be done by something like this:

```powershell
      $Settings = Get-ChildItem -Path '$(System.DefaultWorkingDirectory)' -Filter '*settings.json' -Recurse
      ForEach ($Setting in $Settings) {
          Write-Host "Processing " $Setting.FullName
          (Get-Content -Path $Setting.FullName -Raw) -Replace ',*\r*\n*.*LinterCop.dll"', '' | Out-File $Setting.FullName -Force
      }
```

**Comment by Arthurvdv** [comment]
> I am prepping an example, though ..

@fvet: You should probably wait for @waldo1001 to share a example.

> @waldo1001 Any plans on whether ALOps will support the BCLinterCop on short-term? Or should we revert to custom solutions?
> 
> @jcoe-nvt @Arthurvdv Any script to share ;) ?

Sure! I'm not a PowerShell-guru, so please be gentle about the quality ;-)

```YAML
steps:
- task: PowerShell@2
  displayName: 'Download BusinessCentral.LinterCop.dll'
  inputs:
    targetType: 'inline'
    script: |
      $FileName = "DownloadFile.ps1"
      $GitHubUri = "https://raw.githubusercontent.com/StefanMaron/vsc-lintercop/master/$FileName"
      $BCLinterCopFileName = "BusinessCentral.LinterCop.dll"

      try
      {
          Write-Host "Downloading PowerShell script $GitHubFileName from $GitHubUri..." -ForegroundColor Yellow
          Invoke-WebRequest -Uri $GitHubUri -OutFile $env:AGENT_BUILDDIRECTORY\$FileName
      } 
      catch [System.Net.WebException] 
      {
          Write-Host "Error connecting to $GitHubUri. Please check your file name or repo name and try again." -ForegroundColor Red
      }

      Write-Host "Success. Now executing $FileName" -ForegroundColor Green

      & $env:AGENT_BUILDDIRECTORY\$FileName $env:BUILD_SOURCESDIRECTORY\$BCLinterCopFileName $false
```

**Comment by waldo1001** [comment]
Thanks @Arthurvdv - that is actually very interesting - can I use that in my example?  If so, this could be an example - comments always appreciated!
![image](https://user-images.githubusercontent.com/12088142/165769893-0f72fef6-9524-4675-9da9-8fe14b237436.png)

**Comment by waldo1001** [comment]
The download-part is obviously very dependent on the codecop.. that's why I like to have a powershell step for it in stead of a fixed ALOps step.  Too dependent.

**Comment by waldo1001** [comment]
Here is an example of using custom codecops..

**Comment by dsaveyn** [comment]
We've implemented the example above but still receive the following warnings:

![image](https://user-images.githubusercontent.com/92365904/173060162-592ea614-fd2d-41b7-a65f-fac602385b9c.png)
The (limited) output of the Compile step looks as following:

![image](https://user-images.githubusercontent.com/92365904/173059929-1332e06c-7b70-4335-a7df-bef603455732.png)

However, the Compile Step also shows the correct BCLinterCop Warnings:

![image](https://user-images.githubusercontent.com/92365904/173061221-db7a22c1-044c-4da8-a56b-f4aa38abfdc7.png)

The version of BCLinterCop downloaded was the following (0.25.2 - the latest version):

![image](https://user-images.githubusercontent.com/92365904/173060353-eda0676d-bb23-4751-819e-435c465eb5aa.png)

The VSIX also seems to be the correct one:

![image](https://user-images.githubusercontent.com/92365904/173060592-f1dc87ba-b5e2-4003-9ab3-6c65306d58e7.png)

Does anyone else has the same issue? What am I overlooking?

(I did not upload the Raw Log because it might contains sensitive data).

**Comment by waldo1001** [comment]
Ok, I was able to repro this.
I'll have to talk to development how we would be able to fix this.. .
Thanks for mentioning it!

**Comment by AdminHodor** [comment]
Dear @fvet, @Arthurvdv, @jcoe-nvt,

The fix was released today in ALOps v1.450. Could you please validate?

Kind regards,

**Comment by Arthurvdv** [comment]
Just a heads-up here if somebody is using the example which posted in this thread.
I had a call with @StefanMaron where he updated the download script today, where it needs an (easy) fix to work again.

https://github.com/HodorNV/ALOps/pull/637

The benefit is that the script now downloads the current and the prerelease (vNext) file, where you have to possibility to dynamically switch between these in your pipelines.
